-1

Tested at: 0d4b75b43187040213471d3f7894386a890b1fc6

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-23590/28741/summary.html

I found follow errors while testing this PR

**Failed tests:  Build  ClangBuild**

* **Build**:

I found an error when building: 
 
 <pre>>> Compiling  /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_2_X_2018-06-17-2300/src/RecoParticleFlow/PFClusterTools/src/TreeUtility.cc 
>> Compiling  /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_2_X_2018-06-17-2300/src/RecoParticleFlow/PFClusterTools/src/ClusterClusterMapping.cc 
>> Compiling  /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_2_X_2018-06-17-2300/src/RecoParticleFlow/PFClusterTools/src/PFClusterCalibration.cc 
>> Compiling  /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_2_X_2018-06-17-2300/src/RecoParticleFlow/PFClusterTools/src/CalibCompare.cc 
/build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_2_X_2018-06-17-2300/src/RecoParticleFlow/PFClusterTools/src/PFResolutionMap.cc: In function 'constexpr std::array<float, 18ul> {anonymous}::cPhiGener()':
/build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_2_X_2018-06-17-2300/src/RecoParticleFlow/PFClusterTools/src/PFResolutionMap.cc:228:32: error: missing braces around initializer for 'std::__array_traits<float, 18ul>::_Type {aka float [18]}' [-Werror=missing-braces]
     std::array<float,18> cPhi{0};
                                ^
/build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_2_X_2018-06-17-2300/src/RecoParticleFlow/PFClusterTools/src/PFResolutionMap.cc:229:11: error: call to non-constexpr function 'std::array<_Tp, _Nm>::value_type& std::array<_Tp, _Nm>::operator[](std::array<_Tp, _Nm>::size_type) [with _Tp = float; long unsigned int _Nm = 18ul; std::array<_Tp, _Nm>::reference = float&; std::array<_Tp, _Nm>::value_type = float; std::array<_Tp, _Nm>::size_type = long unsigned int]'
     cPhi[0]=2.97025;
           ^
</pre>

* **Clang**:

I found a compilation error while trying to compile with clang: 
I used this command: 
 scram b vclean && USER_CXXFLAGS='-Wno-register -fsyntax-only' scram build -k -j 16 COMPILER='llvm compile'


 <pre>>> Building LCG reflex dict from header file src/RecoParticleFlow/PFClusterTools/src/classes.h
>> Compiling  /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_2_X_2018-06-17-2300/src/RecoParticleFlow/PFClusterTools/src/PFClusterCalibration.cc 
>> Compiling  /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_2_X_2018-06-17-2300/src/RecoParticleFlow/PFClusterTools/src/CalibCompare.cc 
>> Compiling  /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_2_X_2018-06-17-2300/src/RecoParticleFlow/PFClusterTools/src/ParticleDeposit.cc 
>> Compiling  /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_2_X_2018-06-17-2300/src/RecoParticleFlow/PFClusterTools/src/PFResolutionMapManager.cc 
/build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_2_X_2018-06-17-2300/src/RecoParticleFlow/PFClusterTools/src/PFResolutionMap.cc:227:34: error: constexpr function never produces a constant expression [-Winvalid-constexpr]
  constexpr std::array<float,18> cPhiGener() {
                                 ^
/build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_2_X_2018-06-17-2300/src/RecoParticleFlow/PFClusterTools/src/PFResolutionMap.cc:229:5: note: non-constexpr function 'operator[]' cannot be used in a constant expression
    cPhi[0]=2.97025;
    ^
</pre>

--- a/RecoParticleFlow/PFClusterTools/src/PFResolutionMap.cc
+++ b/RecoParticleFlow/PFClusterTools/src/PFResolutionMap.cc
@@ -19,8 +19,6 @@ PFResolutionMap::PFResolutionMap(const char* name, unsigned nbinseta,
 				 unsigned nbinse, 
 				 double mine, double maxe, double value) 
   : TH2D(name, name, nbinseta, mineta, maxeta, nbinse, mine, maxe) {
-  // fNbinsEta(nbinseta), fMinEta(mineta), fMaxEta(maxeta),
-  //    fNbinsE(nbinse), fMinE(mine), fMaxE(maxe) {
 
   GetXaxis()->SetTitle("#eta");
   GetYaxis()->SetTitle("E");
@@ -37,38 +35,6 @@ PFResolutionMap::PFResolutionMap(const char* name, unsigned nbinseta,
   
 }
 
-// void PFResolutionMap::Init(unsigned nbinseta,  
-// 			   double mineta, double maxeta,
-// 			   unsigned nbinse, 
-// 			   double mine, double maxe) {
-//   assert(mineta<maxeta);
-//   assert(mine<maxe);
-  
-//   unsigned nbins =  nbinseta*nbinse;
-//   fData.reserve( nbins );
-//   for(unsigned i=0; i<nbins; i++) {
-//     fData.push_back(-1);
-//   } 
-  
-//   // calculate lower bound of eta and e bins
-//   double binsize =  (fMaxEta - fMinEta)/fNbinsEta;
-//   double lb = fMinEta;
-//   // cout<<"eta bins lower bounds : "<<endl;
-//   for(unsigned i=0; i<nbinseta; i++) {
-//     fEtaBinsLowerBounds[lb] = i; 
-//     // cout<<i<<" "<<lb<<endl;
-//     lb += binsize;
-//   }
-  
-//   binsize =  (fMaxE - fMinE)/fNbinsE;
-//   lb = fMinE;
-//   // cout<<"E bins lower bounds : "<<endl;
-//   for(unsigned i=0; i<nbinse; i++) {
-//     fEBinsLowerBounds[lb] = i; 
-//     // cout<<i<<" "<<lb<<endl;
-//     lb += binsize;
-//   } 
-// }
 
 PFResolutionMap::PFResolutionMap(const char* name, const char* mapfile) { 
   SetTitle(name);
@@ -138,9 +104,6 @@ bool PFResolutionMap::ReadMapFile(const char* mapfile) {
 
   SetBins(nbinseta, mineta, maxeta, nbinse, mine, maxe);
 
-  // Init(fNbinsEta,fMinEta,fMaxEta,fNbinsE,fMinE,fMaxE);
-
-  // char data[lineSize_];
   char s[lineSize_];
   int pos=inf.tellg();
 
@@ -182,51 +145,6 @@ bool PFResolutionMap::ReadMapFile(const char* mapfile) {
 }
 
 
-
-// int PFResolutionMap::FindBin(double eta, double e) const {
-  
-//   if(eta<fMinEta || eta>=fMaxEta) {
-//     cout<<"PFResolutionMap::FindBin "<<eta<<" out of eta bounds "<<fMinEta<<" "<<fMaxEta<<endl;
-//     return -1;
-//   }
-//   if(e<fMinE || e>=fMaxE) {
-//     cout<<"PFResolutionMap::FindBin "<<e<<" out of e bounds "<<fMinE<<" "<<fMaxE<<endl;
-//     return -1;
-//   }
-  
-//   map<double,unsigned >::const_iterator iteta = 
-//     fEtaBinsLowerBounds.upper_bound( eta );
-//   iteta--;
-		 
-// //   if( iteta != fEtaBinsLowerBounds.end() ) {
-// //     cout<<"eta lower bound "<<iteta->first<<" "<<iteta->second<<endl;
-// //   }
-// //   else return -1;
-
-//   map<double,unsigned>::const_iterator ite = 
-//     fEBinsLowerBounds.upper_bound( e );
-//   ite--;
-// //   if( ite != fEBinsLowerBounds.end() ) {
-// //     cout<<"e lower bound "<<ite->first<<" "<<ite->second<<endl;
-// //   }
-// //   else return -1;
-
-// //   cout<<"returning "<<ite->second * fNbinsEta + iteta->second<<endl;
-// //   cout<<"returning "<<ite->second<<" "<<iteta->second<<endl;
-
-//   return ite->second * fNbinsEta + iteta->second;
-// }
-
-// void PFResolutionMap::Fill(double eta, double e, double res) {
-
-//   unsigned bin = FindBin(eta, e);
-//   if( bin<0 || bin>fData.size() ) {
-//     // cout<<"PFResolutionMap::Fill : out of range " <<bin<<endl;
-//     return;
-//   }
-  
-//   fData[bin] = res;
-// }
 
 
 double PFResolutionMap::getRes(double eta, double phi, double e, int MapEta){
@@ -308,16 +226,17 @@ double PFResolutionMap::minimum(double a,double b){
 double PFResolutionMap::dCrackPhi(double phi, double eta){
 
   constexpr double pi= M_PI;// 3.14159265358979323846;
-  constexpr double twopi= 2.*pi;// 3.14159265358979323846;
-  constexpr double twopiO18= pi/9;// 3.14159265358979323846;
+  constexpr double twopi= 2.*pi;
+  constexpr double twopiO18= pi/9;
   
   //Location of the 18 phi-cracks
-  constexpr std::array<double,18> cPhi {{2.97025, 
-					2.97025-twopiO18, 2.97025-2*twopiO18, 2.97025-3*twopiO18, 2.97025-4*twopiO18,
-					2.97025-5*twopiO18, 2.97025-6*twopiO18, 2.97025-7*twopiO18, 2.97025-8*twopiO18,
-					2.97025-9*twopiO18, 2.97025-10*twopiO18, 2.97025-11*twopiO18, 2.97025-12*twopiO18,
-					2.97025-13*twopiO18, 2.97025-14*twopiO18, 2.97025-15*twopiO18, 2.97025-16*twopiO18,
-					2.97025-17*twopiO18}};
+  constexpr double c0 = 2.97025;
+  constexpr std::array<double,18> cPhi {{c0, 
+					c0-twopiO18, c0-2*twopiO18, c0-3*twopiO18, c0-4*twopiO18,
+					c0-5*twopiO18, c0-6*twopiO18, c0-7*twopiO18, c0-8*twopiO18,
+					c0-9*twopiO18, c0-10*twopiO18, c0-11*twopiO18, c0-12*twopiO18,
+					c0-13*twopiO18, c0-14*twopiO18, c0-15*twopiO18, c0-16*twopiO18,
+					c0-17*twopiO18}};
 
   //Shift of this location if eta<0
   constexpr double delta_cPhi=0.00638;

--- a/RecoParticleFlow/PFClusterTools/src/PFResolutionMap.cc
+++ b/RecoParticleFlow/PFClusterTools/src/PFResolutionMap.cc
@@ -8,6 +8,7 @@
 #include <TMath.h>
 #include <vector>
 using namespace std;
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
 
 
 const unsigned PFResolutionMap::lineSize_ = 10000;
@@ -90,13 +91,13 @@ bool PFResolutionMap::WriteMapFile(const char* mapfile) {
 
   std::ofstream outf(mapfile);
   if( !outf.good() ) {
-    cout<<"PFResolutionMap::Write : cannot open file "<<mapfile<<endl;
+    edm::LogWarning("PFResolutionMap::Write")<<" : cannot open file "<<mapfile;
     return false;
   }
   
   outf<<(*this)<<endl;
   if(!outf.good() ) {
-    cerr<<"PFResolutionMap::Write : corrupted file "<<mapfile<<endl;
+    edm::LogError("PFResolutionMap::Write")<<" : corrupted file "<<mapfile;
     return false;
   }
   else {
@@ -229,10 +230,10 @@ bool PFResolutionMap::ReadMapFile(const char* mapfile) {
 
 
 double PFResolutionMap::getRes(double eta, double phi, double e, int MapEta){
-  static double fMinEta = -2.95;
-  static double fMaxEta = 2.95;
-  static double fMinE=0;
-  static double fMaxE=100;
+  constexpr double fMinEta = -2.95;
+  constexpr double fMaxEta = 2.95;
+  constexpr double fMinE=0;
+  constexpr double fMaxE=100;
 
   if( eta<fMinEta ) eta = fMinEta+0.001;
   if( eta>fMaxEta ) eta = fMaxEta-0.001;
@@ -306,30 +307,34 @@ double PFResolutionMap::minimum(double a,double b){
 //compute the unsigned distance to the closest phi-crack in the barrel
 double PFResolutionMap::dCrackPhi(double phi, double eta){
 
-  static double pi= M_PI;// 3.14159265358979323846;
+  constexpr double pi= M_PI;// 3.14159265358979323846;
+  constexpr double twopi= 2.*pi;// 3.14159265358979323846;
+  constexpr double twopiO18= pi/9;// 3.14159265358979323846;
   
   //Location of the 18 phi-cracks
-  static std::vector<double> cPhi;
-  cPhi.resize(18,0);
-  cPhi[0]=2.97025;
-  for(unsigned i=1;i<=17;i++) cPhi[i]=cPhi[0]-2*i*pi/18;
+  constexpr std::array<double,18> cPhi {2.97025, 
+					2.97025-twopiO18, 2.97025-2*twopiO18, 2.97025-3*twopiO18, 2.97025-4*twopiO18,
+					2.97025-5*twopiO18, 2.97025-6*twopiO18, 2.97025-7*twopiO18, 2.97025-8*twopiO18,
+					2.97025-9*twopiO18, 2.97025-10*twopiO18, 2.97025-11*twopiO18, 2.97025-12*twopiO18,
+					2.97025-13*twopiO18, 2.97025-14*twopiO18, 2.97025-15*twopiO18, 2.97025-16*twopiO18,
+					2.97025-17*twopiO18};
 
   //Shift of this location if eta<0
-  static double delta_cPhi=0.00638;
+  constexpr double delta_cPhi=0.00638;
 
   double m; //the result
 
   //the location is shifted
   if(eta<0){ 
     phi +=delta_cPhi;
-    if(phi>pi) phi-=2*pi;
+    if(phi>pi) phi-=twopi;
   }
   if (phi>=-pi && phi<=pi){
 
     //the problem of the extrema
     if (phi<cPhi[17] || phi>=cPhi[0]){
-      if (phi<0) phi+= 2*pi;
-      m = minimum(phi -cPhi[0],phi-cPhi[17]-2*pi);        	
+      if (phi<0) phi+= twopi;
+      m = minimum(phi -cPhi[0],phi-cPhi[17]-twopi);        	
     }
 
     //between these extrema...
@@ -347,7 +352,7 @@ double PFResolutionMap::dCrackPhi(double phi, double eta){
   }
   else{
     m=0.;        //if there is a problem, we assum that we are in a crack
-    std::cout<<"Problem in dminphi"<<std::endl;
+    edm::LogWarning("PFResolutionMap:Problem")<<"Problem in dminphi";
   }
   if(eta<0) m=-m;   //because of the disymetry
   return m;

--- a/RecoParticleFlow/PFClusterTools/src/PFResolutionMap.cc
+++ b/RecoParticleFlow/PFClusterTools/src/PFResolutionMap.cc
@@ -312,12 +312,12 @@ double PFResolutionMap::dCrackPhi(double phi, double eta){
   constexpr double twopiO18= pi/9;// 3.14159265358979323846;
   
   //Location of the 18 phi-cracks
-  constexpr std::array<double,18> cPhi {2.97025, 
+  constexpr std::array<double,18> cPhi {{2.97025, 
 					2.97025-twopiO18, 2.97025-2*twopiO18, 2.97025-3*twopiO18, 2.97025-4*twopiO18,
 					2.97025-5*twopiO18, 2.97025-6*twopiO18, 2.97025-7*twopiO18, 2.97025-8*twopiO18,
 					2.97025-9*twopiO18, 2.97025-10*twopiO18, 2.97025-11*twopiO18, 2.97025-12*twopiO18,
 					2.97025-13*twopiO18, 2.97025-14*twopiO18, 2.97025-15*twopiO18, 2.97025-16*twopiO18,
-					2.97025-17*twopiO18};
+					2.97025-17*twopiO18}};
 
   //Shift of this location if eta<0
   constexpr double delta_cPhi=0.00638;

